### PR TITLE
Construct equal_exprt in a non-deprecated way

### PR DESCRIPTION
**Pull request by tautschnig**
The default constructor is deprecated.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 49b4c89c277a2bd3c6a647c6b5ae0f7f6948fa57).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97157310

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -2542,8 +2542,8 @@ exprt c_typecheck_baset::do_special_functions(
       throw 0;
     }
 
-    equal_exprt equality_expr;
-    equality_expr.operands()=expr.arguments();
+    equal_exprt equality_expr(
+      expr.arguments().front(), expr.arguments().back());
```
Comment (peterschrammel) [inline src/ansi-c/c_typecheck_expr.cpp:2544]: Are we sure this has two elements?

File: src/solvers/flattening/boolbv_byte_extract.cpp [hunk 2]
```diff
@@ -143,8 +143,8 @@ bvt boolbvt::convert_byte_extract(const byte_extract_exprt &expr)
 
       // add implications
 
-      equal_exprt equality;
-      equality.lhs() = expr.offset(); // index operand
+      // index operand
+      equal_exprt equality(expr.offset(), exprt());
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_byte_extract.cpp]: Add a comment that `rhs` will be set below.

File: src/solvers/flattening/boolbv_byte_extract.cpp [hunk 3]
```diff
@@ -170,8 +170,8 @@ bvt boolbvt::convert_byte_extract(const byte_extract_exprt &expr)
     }
     else
     {
-      equal_exprt equality;
-      equality.lhs() = expr.offset(); // index operand
+      // index operand
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_byte_extract.cpp]: Add a comment that `rhs` will be set below.

File: src/solvers/flattening/boolbv_index.cpp [hunk 4]
```diff
@@ -144,11 +144,10 @@ bvt boolbvt::convert_index(const index_exprt &expr)
 
       // add implications
 
-      equal_exprt index_equality;
-      index_equality.lhs()=index; // index operand
+      // index operand
+      equal_exprt index_equality(index, exprt());
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_index.cpp]: Add a comment that `rhs` will be set below.

File: src/solvers/flattening/boolbv_index.cpp [hunk 5]
```diff
@@ -203,8 +202,8 @@ bvt boolbvt::convert_index(const index_exprt &expr)
...
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_index.cpp]: I'm wondering whether it wouldn't be better to construct the equality on conversion?

**Review by peterschrammel — APPROVED** [review]

File: src/ansi-c/c_typecheck_expr.cpp (hunk 1 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:2544]: That's check 7 lines earlier so this should be ok.

File: src/solvers/flattening/boolbv_index.cpp (hunk 5 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_index.cpp]: I've now done so, for all the cases that you pointed out.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5f0367c0dddfc5af7226d87cf9b2f0209fd6d2a0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97280227

**Pull request merged**